### PR TITLE
cpu: aarch64: optimising memory/thread utilization in BRGEMM Matmul

### PR DESCRIPTION
**Pull request by Shreyas-fuj**
# Description

This PR brings some optimizations to brgemm matmul operator by improving memory utilization and multithreading capabilities.

This PR contains the following changes:

- Modification of blocking parameters for M,K,N based on some heuristics obtained by testing matmul on shapes of majority of language models.
- Assembly level optimization which removes the necessity of the `fadd() `instruction before storing the accumulator results in destination matrix.

## General

- [y] Do all unit and benchdnn tests (`make test` and `make test_benchdnn_*`) pass locally for each commit?
```
make test

99% tests passed, 2 tests failed out of 200

Total Test time (real) = 2142.22 sec

The following tests FAILED:
	159 - test_graph_unit_dnnl_large_partition_usm_cpu (Failed)
	181 - test_benchdnn_modeC_graph_ci_cpu (Failed)
Errors while running CTest
Output from these tests are in: /home/shreyas/G/shr-fuj/oneDNN_open_source/build/Testing/Temporary/LastTest.log
Use "--rerun-failed --output-on-failure" to re-run the failed cases verbosely.
make: *** [Makefile:71: test] Error 8

```
- [y] Have you formatted the code using clang-format?

### CODE REVIEW

File: src/cpu/aarch64/brgemm/brgemm_types.hpp
Comment (dzarukin) [inline src/cpu/aarch64/brgemm/brgemm_types.hpp:195]: `M`, `K`, and `N` are literally declared above as `bcast_dim`, `reduce_dim` and `load_dim`. Any specific reason to add duplicated entries for them?

Comment (Shreyas-fuj) [reply src/cpu/aarch64/brgemm/brgemm_types.hpp:195]: Hi @dzarukin ,
The bcast, reduce and load dimensions are blocked sizes of M,K and N. M, K and N are actual sizes of the original matrix.

File: .gitignore
Comment (mgouicem) [inline .gitignore]: If it helps, you can use build* so that any directory starting with `build` prefix would be ignored.

**Comment by theComputeKid** [comment]
@Sqvid can you please review this for the aarch64? Thanks.

**Comment by Sqvid** [comment]
@Shreyas-fuj Thanks for your contribution. In addition to some of the comments could you also squash and re-push with a commit message that follows the conventions outlined here: https://github.com/oneapi-src/oneDNN/blob/main/CONTRIBUTING.md#code-contribution-guidelines. Thanks!

**Comment by vpirogov** [comment]
We need a commit message linter...

File: .gitignore
Comment (Shreyas-fuj) [reply .gitignore]: Sure @mgouicem, thanks. But I have restored gitignore file in the next commit.

**Comment by Shreyas-fuj** [comment]
Hi @Sqvid ,
I have squashed the commits in to a single one by following the guidelines. Thanks.

File: src/cpu/aarch64/brgemm/brgemm_types.hpp
Comment (mgouicem) [inline src/cpu/aarch64/brgemm/brgemm_types.hpp:195]: Are those used anywhere? I see them set in brgemm_matmul.cpp, but it seems only M,N,K from brgemm_matmul_conf or matmul_params are used all over (and not from brgemm_t).

**Review by mgouicem — COMMENTED** [review]
> Hi @Sqvid , I have squashed the commits in to a single one by following the guidelines. Thanks.

Thanks. (nit) commit message: `hueristics -> heuristics`.

File: src/cpu/aarch64/brgemm/jit_brgemm_kernel.cpp
Comment (Shreyas-fuj) [inline src/cpu/aarch64/brgemm/jit_brgemm_kernel.cpp:775]: @mgouicem, K has been used in here, M,N is put since it would also be used in future when improving the sve matmuls.

Comment (mgouicem) [reply src/cpu/aarch64/brgemm/jit_brgemm_kernel.cpp:775]: Thanks for pointing out the location. I guess I have 2 questions:
- First is API related: The main issue I see with putting global M/N/K in brgemm_t is that it will lead to move some of the driver logic (so the code that configures brgemm object), directly to brgemm_t class. By doing so, it might make configuring brgemm objects harder for driver code. Hence, why not just rely on apply_beta? (also curious why `beta = 1 && K=reduce_dim` should be treated differently than `beta = 1 && K != reduce_dim` from brgemm perspective).
- Second is performnace related: in general it is beneficial to start with 0 accumulators, and do the fadd at the end upon loading for better cache locality, as loadC -> fadd -> storeC happen in a short window of time. Could you share some performance data highlighting the impact of your changes?

File: src/cpu/aarch64/brgemm/brgemm_types.hpp
Comment (mgouicem) [reply src/cpu/aarch64/brgemm/brgemm_types.hpp:195]: continuing discussion [here](https://github.com/oneapi-src/oneDNN/pull/2103/files#r1772962084), thanks.

**Comment by theComputeKid** [comment]
@vpirogov we can have a python script called by a dedicated github action, that fails a PR if the start of the PR does not start with a word followed by a semi colon. I can help. Thoughts?

**Comment by vpirogov** [comment]
> @vpirogov we can have a python script called by a dedicated github action, that fails a PR if the start of the PR does not start with a word followed by a semi colon. I can help with this, we can create an issue and I can work on it if you think it is worth it. Thoughts?

Sounds like a plan. In ideal world I would use an off the shelf action for this, but I don't see any. The only tool I was able to find is [CommitLint](https://github.com/conventional-changelog/commitlint).

File: src/cpu/aarch64/brgemm/jit_brgemm_kernel.cpp
Comment (Shreyas-fuj) [reply src/cpu/aarch64/brgemm/jit_brgemm_kernel.cpp:775]: 1. In this PR we have not changed any alpha beta logic, the functionality is currently moved from apply_alpha_beta function to zero_accumulators function by keeping the previous logic intact. Now according to previous logic loadC -> fadd -> storeC was happening even when reduce_dim == K. Now when reduce_dim == K, the result alsong the entire K dimension is calculated at once, so the previous destination values will anyway be zero. So there is no need to apply beta when K=reduce_dim(since n * 0 = 0). Therefore to maintain previous logic we have kept the condition need_to_apply_beta, and for the above mentioned reason we have added K != reduce_dim.

2. First point here is as mentioned above if reduce_dim = K, the entire operation loadC -> fadd -> storeC  can be avoided, if not then instead of zeroing the registers we are loading the previous destination values. Taking performance into account, no.of times zero_accumulators function is called is same as no.of times apply_alpha_beta function is called. So just wanted to know how it would affect if we load before multiplication vs load after multiplication. Since we are loading here before multiplication, the JIT code generated reduces.

3. Below I have pasted the performance results of the matmuls run on shapes of multiple highly used models. All results are obtained after all the changes mentioned in this commit.
Machine : GR3E
Cores : 32
Time in : ms
Ratio above 1 means the new implementation is better. 
Ration within 0.95-1.05 can be ignored taking into account run to run variation.
All timings are recorder with performance mode in benchdnn.

<html xmlns:v="urn:schemas-microsoft-com:vml"
xmlns:o="urn:schemas-microsoft-com:office:office"
xmlns:x="urn:schemas-microsoft-com:office:excel"
xmlns="http://www.w3.org/TR/REC-html40">

<head>

<meta name=ProgId content=Excel.Sheet>
<meta name=Generator content="Microsoft Excel 15">
<link id=Main-File rel=Main-File
href="file:///C:/Users/SHANKA~1/AppData/Local/Temp/msohtmlclip1/01/clip.htm">
<link rel=File-List
href="file:///C:/Users/SHANKA~1/AppData/Local/Temp/msohtmlclip1/01/clip_filelist.xml">
<style>

</style>
</head>

<body link="#0563C1" vlink="#954F72">


shapes | model | sve_old | sve_new | Speed up
-- | -- | -- | -- | --
4096x4096:4096x128256 | llama_2_ | 1766.74 | 1752.73 | 1.007993
2048x4096:4096x128256 | llama_2_ | 880.027 | 868.45 | 1.013331
580x4096:4096x128256 | llama_2_ | 260.519 | 249.232 | 1.045287
4096x4096:4096x14336 | llama_2_ | 198.13 | 196.327 | 1.009184
4096x14336:14336x4096 | llama_2_ | 198.125 | 195.018 | 1.015932
290x4096:4096x128256 | llama_2_ | 140.4 | 124.221 | 1.130244
2048x14336:14336x4096 | llama_2_ | 98.4778 | 97.3108 | 1.011993
2048x4096:4096x14336 | llama_2_ | 98.4486 | 97.5129 | 1.009596
145x4096:4096x128256 | llama_2_ | 75.2273 | 62.0745 | 1.211887
4096x4096:4096x4096 | llama_2_ | 56.7314 | 56.0889 | 1.011455
2048x4096:4096x4096 | llama_2_ | 28.2013 | 28.0062 | 1.006966
580x14336:14336x4096 | llama_2_ | 28.7445 | 27.8955 | 1.030435
580x4096:4096x14336 | llama_2_ | 27.985 | 27.6562 | 1.011889
4096x4096:4096x1024 | llama_2_ | 14.1047 | 14.4625 | 0.97526
290x4096:4096x14336 | llama_2_ | 14.1301 | 13.8063 | 1.023453
290x14336:14336x4096 | llama_2_ | 14.0387 | 13.7204 | 1.023199
580x4096:4096x4096 | llama_2_ | 8.12461 | 8.05997 | 1.00802
2048x4096:4096x1024 | llama_2_ | 7.05128 | 7.1036 | 0.992635
145x4096:4096x14336 | llama_2_ | 7.30387 | 6.97038 | 1.047844
145x14336:14336x4096 | llama_2_ | 7.24987 | 6.90561 | 1.049852
4096x16:16x128256 | llama_2_ | 11.0101 | 11.3132 | 0.973208
290x4096:4096x4096 | llama_2_ | 4.09645 | 3.94733 | 1.037777
2048x16:16x128256 | llama_2_ | 5.46918 | 5.74414 | 0.952132
145x4096:4096x4096 | llama_2_ | 2.08148 | 1.98685 | 1.047628
580x4096:4096x1024 | llama_2_ | 2.04714 | 2.05082 | 0.998206
290x4096:4096x1024 | llama_2_ | 1.04004 | 1.00824 | 1.03154
580x16:16x128256 | llama_2_ | 1.79656 | 1.63177 | 1.100988
4096x14336:14336x16 | llama_2_ | 1.23134 | 1.14104 | 1.079138
4096x16:16x14336 | llama_2_ | 1.18164 | 1.23273 | 0.958555
145x4096:4096x1024 | llama_2_ | 0.552641 | 0.519076 | 1.064663
290x16:16x128256 | llama_2_ | 0.90235 | 0.742302 | 1.21561
2048x16:16x14336 | llama_2_ | 0.510833 | 0.50407 | 1.013417
2048x14336:14336x16 | llama_2_ | 0.657688 | 0.548903 | 1.198186
145x16:16x128256 | llama_2_ | 0.317327 | 0.288596 | 1.099554
4096x4096:4096x16 | llama_2_ | 0.37306 | 0.376977 | 0.989609
4096x16:16x4096 | llama_2_ | 0.265271 | 0.245874 | 1.07889
2048x16:16x4096 | llama_2_ | 0.142499 | 0.122595 | 1.162356
2048x4096:4096x16 | llama_2_ | 0.187398 | 0.142864 | 1.311723
580x16:16x14336 | llama_2_ | 0.132808 | 0.123195 | 1.078031
580x14336:14336x16 | llama_2_ | 0.198657 | 0.161385 | 1.230951
4096x16:16x1024 | llama_2_ | 0.083001 | 0.062224 | 1.333901
290x14336:14336x16 | llama_2_ | 0.162207 | 0.079482 | 2.040807
290x16:16x14336 | llama_2_ | 0.073193 | 0.062326 | 1.174358
580x4096:4096x16 | llama_2_ | 0.049859 | 0.039704 | 1.255768
580x16:16x4096 | llama_2_ | 0.048042 | 0.037274 | 1.288886
2048x16:16x1024 | llama_2_ | 0.05377 | 0.033017 | 1.628553
145x14336:14336x16 | llama_2_ | 0.180858 | 0.045537 | 3.971645
145x16:16x14336 | llama_2_ | 0.043121 | 0.032735 | 1.317291
290x4096:4096x16 | llama_2_ | 0.041729 | 0.025885 | 1.612125
290x16:16x4096 | llama_2_ | 0.031354 | 0.020774 | 1.509259
580x16:16x1024 | llama_2_ | 0.023247 | 0.012016 | 1.934752
145x4096:4096x16 | llama_2_ | 0.044779 | 0.01392 | 3.21696
145x16:16x4096 | llama_2_ | 0.022643 | 0.01254 | 1.805751
290x16:16x1024 | llama_2_ | 0.018561 | 0.008141 | 2.280041
145x16:16x1024 | llama_2_ | 0.010585 | 0.006031 | 1.755266
45x5120:5120x5120 | llama, | 1.64412 | 1.00751 | 1.631865
40x45x128:40x128x45 | llama, | 0.021532 | 0.016753 | 1.28526
40x45x45:40x45x128 | llama, | 0.012503 | 0.013054 | 0.957753
45x5120:5120x13824 | llama, | 3.38549 | 2.84688 | 1.189193
45x13824:13824x5120 | llama, | 4.53799 | 2.73022 | 1.662133
45x5120:5120x32000 | llama, | 7.75908 | 6.45042 | 1.20288
12x41x64:12x64x41 | distil-gpt2, | 0.006887 | 0.005359 | 1.28521
12x41x41:12x41x64 | distil-gpt2, | 0.004796 | 0.005873 | 0.816648
41x768:768x50257 | distil-gpt2, | 1.53154 | 1.508 | 1.01561
68x4096:4096x4096 | intel_nueral, | 1.05818 | 0.945752 | 1.118877
68x4096:4096x1024 | intel_nueral, | 0.275688 | 0.245584 | 1.122581
68x4096:4096x14336 | intel_nueral, | 3.77038 | 3.36089 | 1.12184
68x14336:14336x4096 | intel_nueral, | 3.74343 | 3.31704 | 1.128545
68x4096:4096x32000 | intel_nueral, | 9.11819 | 7.56396 | 1.205478
89x5120:5120x5120 | phi, | 2.16478 | 1.92083 | 1.127002
40x89x128:40x128x89 | phi, | 0.042452 | 0.044692 | 0.949871
40x89x89:40x89x128 | phi, | 0.03543 | 0.035462 | 0.99912
89x5120:5120x13824 | phi, | 5.80715 | 5.38709 | 1.077975
89x13824:13824x5120 | phi, | 5.63184 | 5.19504 | 1.08408
89x5120:5120x32003 | phi, | 17.1893 | 12.2598 | 1.402086
12x53x64:12x64x53 | tinyllama, | 0.006315 | 0.00616 | 1.025054
12x53x53:12x53x64 | tinyllama, | 0.005899 | 0.005899 | 1.000002
53x768:768x50257 | tinyllama, | 2.23117 | 1.88322 | 1.184763
8x850x32:8x32x850 | detr, | 0.16123 | 0.159554 | 1.010504
8x850x850:8x850x32 | detr, | 0.146462 | 0.145891 | 1.003914
850x256:256x256 | detr, | 0.055143 | 0.046073 | 1.196877
850x256:256x2048 | detr, | 0.364014 | 0.347224 | 1.048355
850x2048:2048x256 | detr, | 0.373092 | 0.34962 | 1.067136
100x256:256x256 | detr, | 0.011366 | 0.008818 | 1.288941
8x100x32:8x32x100 | detr, | 0.007287 | 0.006218 | 1.171803
8x100x100:8x100x32 | detr, | 0.005892 | 0.005648 | 1.043108
8x100x32:8x32x850 | detr, | 0.022562 | 0.02285 | 0.987392
8x100x850:8x850x32 | detr, | 0.020517 | 0.021184 | 0.968486
100x256:256x2048 | detr, | 0.051415 | 0.044776 | 1.148262
100x2048:2048x256 | detr, | 0.058018 | 0.045518 | 1.274626
100x256:256x92 | detr, | 0.009906 | 0.006462 | 1.532928
512x768:768x768 | bert,hf_bart,hf_longformer,hf_roberta_base,hf_albert | 0.242001 | 0.237409 | 1.019342
12x512x64:12x64x512 | bert,hf_longformer,hf_roberta_base,hf_albert | 0.182308 | 0.162609 | 1.121143
12x512x512:12x512x64 | bert,hf_roberta_base,hf_albert | 0.173427 | 0.159206 | 1.089325
512x768:768x3072 | bert,hf_bart,hf_longformer,hf_roberta_base,hf_albert | 0.978512 | 0.952415 | 1.027401
512x3072:3072x768 | bert,hf_bart,hf_longformer,hf_roberta_base,hf_albert | 1.0788 | 1.00064 | 1.07811
197x768:768x768 | vit, | 0.097467 | 0.097011 | 1.004695
12x197x64:12x64x197 | vit, | 0.034134 | 0.029575 | 1.154121
12x197x197:12x197x64 | vit, | 0.027059 | 0.027354 | 0.989208
197x768:768x3072 | vit, | 0.374175 | 0.376631 | 0.993479
197x3072:3072x768 | vit, | 0.50158 | 0.385823 | 1.300026
12x256x64:12x64x256 | hf_gpt2 | 0.052364 | 0.043284 | 1.20978
12x256x256:12x256x64 | hf_gpt2 | 0.050892 | 0.043206 | 1.177902
256x768:768x50257 | hf_gpt2 | 9.31642 | 8.3162 | 1.120274
192x512:512x2048 | hf_clip | 0.172951 | 0.162962 | 1.061296
384x50x50:384x50x64 | hf_clip | 0.053128 | 0.053656 | 0.990158
32x512:512x512 | hf_clip | 0.01016 | 0.010751 | 0.944991
32x512:512x32 | hf_clip | 0.009652 | 0.005385 | 1.792263
1600x3072:3072x768 | hf_clip | 3.11312 | 3.05703 | 1.018348
192x2048:2048x512 | hf_clip | 0.163615 | 0.16493 | 0.992027
32x768:768x512 | hf_clip | 0.013652 | 0.014225 | 0.959719
192x512:512x512 | hf_clip | 0.045665 | 0.043849 | 1.041422
1600x768:768x3072 | hf_clip | 3.03149 | 3.02295 | 1.002825
384x50x64:384x64x50 | hf_clip | 0.069855 | 0.067996 | 1.027338
1600x768:768x768 | hf_clip | 0.771308 | 0.747592 | 1.031723
512x768:768x50265 | hf_bart,hf_longformer | 17.1542 | 15.7876 | 1.086562
8x512x512:8x512x64 | hf_t5 | 0.116126 | 0.107556 | 1.079679
512x2048:2048x512 | hf_t5 | 0.451532 | 0.420167 | 1.074649
8x512x64:8x64x512 | hf_t5 | 0.128603 | 0.109472 | 1.174757
512x512:512x512 | hf_t5 | 0.117158 | 0.107047 | 1.094454
512x512:512x32128 | hf_t5 | 7.50572 | 6.86368 | 1.093542
512x512:512x2048 | hf_t5 | 0.438574 | 0.41989 | 1.044497
512x3072:3072x1024 | hf_bert_large | 1.3301 | 1.31719 | 1.009801
16x512x512:16x512x64 | hf_bert_large,hf_t5_large | 0.231442 | 0.211797 | 1.092754
512x1024:1024x3072 | hf_bert_large | 1.31294 | 1.28848 | 1.018984
512x1024:1024x30522 | hf_bert_large | 14.2754 | 12.8995 | 1.106663
512x1024:1024x1024 | hf_bert_large,hf_t5_large | 0.43552 | 0.416528 | 1.045596
16x512x64:16x64x512 | hf_bert_large,hf_t5_large | 0.235142 | 0.215554 | 1.090873
24x256x768:24x768x64 | hf_longformer | 0.250404 | 0.244617 | 1.023657
256x32x33:256x33x64 | llama | 0.018385 | 0.019202 | 0.957443
32x512:512x32000 | llama | 0.540057 | 0.505382 | 1.068611
1024x512:512x512 | llama | 0.23153 | 0.211223 | 1.09614
1024x1536:1536x512 | llama | 0.671995 | 0.642275 | 1.046273
256x32x64:256x64x33 | llama | 0.024032 | 0.02436 | 0.986511
1024x512:512x1536 | llama | 0.651934 | 0.629817 | 1.035117
1024x1280:1280x50257 | hf_gpt2_large | 53.936 | 54.6074 | 0.987705
20x1024x64:20x64x1024 | hf_gpt2_large | 1.15325 | 1.10585 | 1.042863
20x1024x1024:20x1024x64 | hf_gpt2_large | 1.09812 | 1.0864 | 1.010788
832x768:768x3072 | hf_big_bird | 1.59263 | 1.57681 | 1.010033
108x64x64:108x64x192 | hf_big_bird | 0.075833 | 0.073341 | 1.033976
832x768:768x768 | hf_big_bird | 0.394577 | 0.381855 | 1.033316
12x64x448:12x448x64 | hf_big_bird | 0.033897 | 0.021844 | 1.551825
832x3072:3072x768 | hf_big_bird | 1.6938 | 1.61595 | 1.048176
819x768:768x50358 | hf_big_bird | 25.6993 | 25.5751 | 1.004856
12x64x64:12x64x448 | hf_big_bird | 0.023374 | 0.023843 | 0.98033
108x64x192:108x192x64 | hf_big_bird | 0.07171 | 0.07339 | 0.977102
819x768:768x768 | hf_big_bird | 0.393133 | 0.379265 | 1.036565
12x64x64:12x64x832 | hf_big_bird | 0.036657 | 0.03708 | 0.98859
12x64x832:12x832x64 | hf_big_bird | 0.060499 | 0.036494 | 1.657792
12x576x64:12x64x64 | hf_big_bird | 0.043469 | 0.026585 | 1.635122
12000x384:384x1536 | hf_whisper | 5.60737 | 5.63077 | 0.995844
12000x384:384x384 | hf_whisper | 1.42838 | 1.3842 | 1.031917
12000x1536:1536x384 | hf_whisper | 5.8769 | 5.8006 | 1.013154
12000x384:384x256 | hf_whisper | 0.963903 | 0.922938 | 1.044385
512x768:768x250002 | hf_roberta_base | 87.9058 | 77.7651 | 1.130402
512x768:768x128 | hf_albert | 0.04556 | 0.042705 | 1.066849
512x128:128x768 | hf_albert | 0.048218 | 0.043462 | 1.109422
512x128:128x30000 | hf_albert | 1.60461 | 1.59525 | 1.005867
2048x256:256x768 | hf_reformer | 0.336265 | 0.312458 | 1.076193
2048x512:512x256 | hf_reformer | 0.231212 | 0.211152 | 1.095003
2048x512:512x320 | hf_reformer | 0.270565 | 0.262268 | 1.031636
384x64x64:384x64x128 | hf_reformer | 0.167904 | 0.165479 | 1.014654
12x2048x64:12x64x64 | hf_reformer | 0.144387 | 0.082711 | 1.745672
2048x768:768x256 | hf_reformer | 0.334619 | 0.313146 | 1.068572
384x64x128:384x128x64 | hf_reformer | 0.166376 | 0.163681 | 1.016465
2048x256:256x512 | hf_reformer | 0.230792 | 0.209442 | 1.101938
512x1024:1024x32128 | hf_t5_large | 14.5651 | 13.6375 | 1.068018
512x4096:4096x1024 | hf_t5_large | 1.77372 | 1.79337 | 0.989043
512x1024:1024x4096 | hf_t5_large | 1.73756 | 1.7108 | 1.015642
6x3401x3401:6x3401x64 | yolo | 3.74399 | 3.73841 | 1.001493
3401x1536:1536x384 | yolo | 1.6592 | 1.59497 | 1.04027
3401x384:384x384 | yolo | 0.405752 | 0.389557 | 1.041573
6x3401x64:6x64x3401 | yolo | 3.75784 | 3.87955 | 0.968628
3401x384:384x1536 | yolo | 1.71531 | 1.55558 | 1.102682



</body>

</html>

**Comment by Sqvid** [comment]
> The only tool I was able to find is [CommitLint](https://github.com/conventional-changelog/commitlint)

CommitLint looks very capable for the task. The only issue I could see (which is not necessarily a bad thing) is that it allows the sub-category to be in parenthesis `cpu(aarch64):` rather than consecutive colons `cpu: aarch64:`. However, it's not a terrible thing if we were to follow a standard like [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0/).

Either way a custom parser should be fairly easy to throw together with some regex.

File: src/cpu/aarch64/brgemm/jit_brgemm_kernel.cpp
Comment (mgouicem) [reply src/cpu/aarch64/brgemm/jit_brgemm_kernel.cpp:775]: Thanks for the performance data. I assume this is with matmul driver and no post-ops right?
Some of the results are unexpected, in particular for cases with small K, I would think that `K == reduce_dim` in those cases, which would generate the same code with and without patch, yet we see some speedups there too. For example:

| shapes | model | sve_od | sve_new | speedup up |
| --- | --- | --- | --- | --- |
| 2048x16:16x1024 | llama_2_ | 0.05377  | 0.033017 | 1.628553
| 580x16:16x1024  | llama_2_ | 0.023247 | 0.012016 | 1.934752
| 290x16:16x1024  | llama_2_ | 0.018561 | 0.008141 | 2.280041

Would you have an explanation for those?

> Since we are loading here before multiplication, the JIT code generated reduces.

Ok so if I get it right, the speedup comes from code size reduction (+ fewer fadds when beta == 1)?

> First point here is as mentioned above if reduce_dim = K, the entire operation loadC -> fadd -> storeC can be avoided, if not then instead of zeroing the registers we are loading the previous destination values.

The point I am trying to make is that, for matmul without post-op, your optimization seems more related to `beta != 0`, than `reduce_dim == K`.
Indeed, the optimization you mention is valid for both `beta != 0 && K == reduce_dim` and `beta != 0 && K != reduce_dim`. From matmul perspective, 
- if K-blocking happens, brgemm beta is set to 1.f for accumulation between K-blocks.
- if no K-blocking happens, then beta is set to 0 if no post-ops, or it is set to 1.0f if there is a sum post-op.

If the optimization is independent of M,N,K global shapes, then we can keep brgemm independant of those and maintain separation of concerns (matmul blocking logic does not leak to brgemm).

Comment (Shreyas-fuj) [reply src/cpu/aarch64/brgemm/jit_brgemm_kernel.cpp:775]: - Firstly regarding the speed up, in the cases you have highlighted the speed up comes from the fact the N_blk is reduced to 32 from 64, when M,K are small.
- Secondly the assembly optimization(fadd) only reduces the JIT code, but no much change in performance is observed. But we feel its a right direction to proceed in optimizing the JIT code generated. 
- Now regarding regarding alpha and beta, they are set here : https://github.com/oneapi-src/oneDNN/blob/main/src/cpu/aarch64/matmul/brgemm_matmul.cpp#L134 . Here they are not setting it based on relation between K and reduce_dim. So the condition when `K == reduce_dim` there is no need of fadd what ever beta may be. I have observed this extra fadds even when K is not blocked and without postops. Therefore for me to inculcate this in beta itself, would require some more understanding in this, so I have kept the current logic intact and added the condition `K == reduce_dim` . May be I can remember this and apply this in the next set of PRs for matmul.
- Currently my understanding is the function apply_alpha_beta would be called irrespective of post-ops.

Comment (mgouicem) [reply src/cpu/aarch64/brgemm/jit_brgemm_kernel.cpp:775]: > Secondly the assembly optimization(fadd) only reduces the JIT code, but no much change in performance is observed. But we feel its a right direction to proceed in optimizing the JIT code generated. But still, if you feel that cache utilization is better in the previous approach, please throw some light on that.

I am not advocating that one is better than the other for performance reasons.
However, if this does not improve performance, I would advocate to keep M, N, K (or any logic related to Matmul) out of brgemm for the following reasons:
- brgemm can be used in different scope where M/N/K do not make much sense (e.g. convolution)
- if brgemm kernel jitting depends on caller information, it becomes harder to reuse this code for external ukernel API ([API](https://github.com/oneapi-src/oneDNN/blob/main/include/oneapi/dnnl/dnnl_ukernel.hpp), [documentation](https://oneapi-src.github.io/oneDNN/dev_guide_ukernel_basic_concepts.html)).

I would let aarch64 maintainers masters of this decision though.
@theComputeKid @Sqvid

Comment (Shreyas-fuj) [reply src/cpu/aarch64/brgemm/jit_brgemm_kernel.cpp]: > I am not advocating that one is better than the other for performance reasons.
> However, if this does not improve performance, I would advocate to keep M, N, K (or any logic related to Matmul) out of brgemm for the following reasons:
> 
> brgemm can be used in different scope where M/N/K do not make much sense (e.g. convolution)

@mgouicem , I agree with you on this point. For now I have removed the global variables M,K,N from brgemm. Thanks.

Comment (mgouicem) [reply src/cpu/aarch64/brgemm/jit_brgemm_kernel.cpp]: THanks!

**Comment by Shreyas-fuj** [comment]
@mgouicem,  @dzarukin, please let me know if any more changes are required for approval, thanks.

**Comment by Shreyas-fuj** [comment]
@mgouicem @dzarukin, wanted to know if anything from my side is remaining for approval of the PR, thanks.

**Comment by mgouicem** [comment]
Hi @Shreyas-fuj. This looks fine on my side.
Given the aarch64 target, you will need approval from @oneapi-src/onednn-cpu-aarch64  code owners.

**Review by mgouicem — APPROVED** [review]

**Comment by Shreyas-fuj** [comment]
@theComputeKid, @theComputeKid requesting for review and approval of the PR, thanks.

**Comment by Sqvid** [comment]
@Shreyas-fuj Yes looking into now. Will revert in ~48 hours. Thanks.

**Comment by Sqvid** [comment]
@Shreyas-fuj performance looks good to me.

Just a couple of nitpicks I'd appreciate if you could address:
    1. Two files are no longer needed in the patchset (as the changes have been removed).
    2. In some cases you have deleted or moved code and left a comment in the original location to the effect of "deleted this.". This makes sense in a side-by-side diff but will be confusing to a future reader.

Other than this I'm happy to approve the changes.

**Comment by Shreyas-fuj** [comment]
> @Shreyas-fuj performance looks good to me.
> 
> Just a couple of nitpicks I'd appreciate if you could address: 1. Two files are no longer needed in the patchset (as the changes have been removed). 2. In some cases you have deleted or moved code and left a comment in the original location to the effect of "deleted this.". This makes sense in a side-by-side diff but will be confusing to a future reader.
> 
> Other than this I'm happy to approve the changes. Thanks.

@Sqvid , I have applied the suggested changes, please have a look, thanks.

**Review by Radu2k — APPROVED** [review]
LGTM.

**Comment by spalicki** [comment]
@Shreyas-fuj Could you please rebase the changes on top of main and squash the commits? Or change the PR name to match the desired commit title. If you could do both it would be even better - merging PR with multiple commits squashes the PR (we want to avoid merge commits on release branches) and takes the commit message from PR title, which currently is too long and not in sync with our guidelines.

**Comment by Shreyas-fuj** [comment]
> @Shreyas-fuj Could you please rebase the changes on top of main and squash the commits? Or change the PR name to match the desired commit title. If you could do both it would be even better - merging PR with multiple commits squashes the PR (we want to avoid merge commits on release branches) and takes the commit message from PR title, which currently is too long and not in sync with our guidelines.

@spalicki , I have done both rebasing and squashing of the commits as suggested, thanks.

**Pull request merged**
